### PR TITLE
clean up and refactor build

### LINKED ISSUE #5209: Stop building noarch conda packages
The status of no-arch packages seems unclear, and continuing to use them seems risky.  A new version of conda-build caused a release candidate build to partially fail (only after several hours and in the middle of the night, of course).

### LINKED ISSUE #5679: Deploy script should automatically create and upload examples tarballs
Currently done manually on full releases.

### PR DESCRIPTION
**Pull request by bryevdv**
issues: closes #5209 closes #5679

* Refactor the `setup.py` nightmare into simple `setup.py` and `_setup_support.py`
* Update to current tool versions: `conda=4.3.4`, `conda-env=2.6.0`,  `conda-build=2.1.1`
* Adhere to PEP 440 versions:
  - `0.12.5`
  - `0.12.5.dev2`
  - `0.12.5rc1`
  and on travis, local versions like `bokeh-0.12.4+20.gddad8b0-py27`

* remove all `__conda_version__` and `TRAVIS_BUILD_NUMBER` junk
* remove `get_bump_version.py`
* simplify `meta.yaml` with jinja templates
* store classifiers in sidecar file `classifiers.txt`

**HERE is the change that merits discussion:**

* use `setuptools` in ALL cases now

This allows cutting out a special case for windows, removing out custom `develop` and `find_packages` code, and handling entry points correctly. There were some reasons ~5 years ago not to use setuptools but I think that it's now OK. 

ping @mattpap @damianavila @bokeh/dev 

Still to do:
* [x] replace `noarch` packages for: `flexx`, `pytest-base-url`, and `pytest-selenium`
* [x] rewrite `build_and_upload.sh` as a python script with detailed status and error checks
  - [x] use `conda convert` to create non-`noarch` packages
  - [x] update upload for non-`noarch` packages
  - [x] create and upload examples tarball 

We do need to verify that Bokeh works in venv's still but AFAICT it should now be fine still.

### CODE REVIEW

**Comment by bryevdv** [comment]
New complete `setup.py` 

```
#-----------------------------------------------------------------------------
# Copyright (c) 2012 - 2017, Continuum Analytics, Inc. All rights reserved.
#
# Powered by the Bokeh Development Team.
#
# The full license is in the file LICENCE.txt, distributed with this software.
#-----------------------------------------------------------------------------
''' Setup script for Bokeh.
'''
from os.path import join
from shutil import copy
import sys

from setuptools import find_packages, setup

from _setup_support import (build_or_install_bokehjs, clean, develop,
                            fixup_argv_for_sdist, fixup_old_js_args, install,
                            get_cmdclass, get_package_data, get_version,
                            install_js, package_files, package_path,
                            ROOT, SERVER, show_help)

# immediately bail for ancient pythons
if sys.version_info[:2] < (2, 7):
    raise RuntimeError("Bokeh requires python >= 2.7")

# immediately handle lightweight "python setup.py --install-js"
if len(sys.argv) == 2 and sys.argv[-1] == '--install-js':
    install_js()
    sys.exit()

# we want to have the license at the top level of the GitHub repo, but setup
# can't include it from there, so copy it to the package directory first thing
copy("LICENSE.txt", "bokeh/")

# state our install requirements here
# TODO (bev) single source of truth? Should this come from meta.yaml?
REQUIRES = [
    'six>=1.5.2',
    'requests>=1.2.3',
    'PyYAML>=3.10',
    'python-dateutil>=2.1',
    'Jinja2>=2.7',
    'numpy>=1.7.1',
    'tornado>=4.3',
]

# handle the compatibility difference for futures package for python 2
if sys.version_info[:2] == (2, 7):
    REQUIRES.append('futures>=3.0.3')

fixup_old_js_args()

fixup_argv_for_sdist()

jsbuild, jsinstall = build_or_install_bokehjs()

# configuration to include all the special or non-python files in the package
# directory that need to also be installed or included in a build
sampledata_pats = ('.csv', '.conf', '.gz', '.json', '.png', '.ics', '.geojson')
package_path(join(SERVER, 'static'))
package_path(join(ROOT, 'bokeh', 'core', '_templates'))
package_path(join(ROOT, 'bokeh', 'sphinxext', '_templates'))
package_path(join(ROOT, 'bokeh', 'server', 'views'), ('.html'))
package_path(join(ROOT, 'bokeh', 'sampledata'), sampledata_pats)
package_files('LICENSE.txt', 'themes/*.yaml')

setup(

    # basic package metadata
    name='bokeh',
    version=get_version(),
    description='Interactive plots and applications in the browser from Python',
    license='New BSD',
    author='Continuum Analytics',
    author_email='info@continuum.io',
    url='http://github.com/bokeh/bokeh',
    classifiers=open("classifiers.txt").read().split('\n'),
    install_requires=REQUIRES,

    # details needed by setup
    packages=find_packages(),
    package_data=get_package_data(),
    entry_points={'console_scripts': ['bokeh = bokeh.__main__:main',], },
    zip_safe=False,
    cmdclass=get_cmdclass()

)

# report on
if 'clean'   in sys.argv: clean()
if '--help'  in sys.argv: show_help(jsbuild, jsinstall)
if 'develop' in sys.argv: develop(jsbuild, jsinstall)
if 'install' in sys.argv: install(jsbuild, jsinstall)
```

**Comment by mattpap** [comment]
>  TODO (bev) single source of truth? Should this come from meta.yaml?

It should, but that would mean to depend on `yaml` in setup, though getting dependencies manually with a little code isn't that hard. This would simplify releases, because one wouldn't need anymore to verify those deps.

**Comment by bryevdv** [comment]
> It should, but that would mean to depend on yaml in setup, though getting dependencies manually with a little code isn't that hard. This would simplify releases, because one wouldn't need anymore to verify those deps.

Now that `meta.yaml` is a jinja template, I think we'd have to use `conda_build.metadata` since it's not a valid yaml file anymore.

**Comment by bryevdv** [comment]
Something like this is another option but I am not sure if it handles `futures` correctly or not:
```
  run:
    {{ data.get('install_requires') }}
```

**Comment by bryevdv** [comment]
@mattpap @birdsarah @damianavila @bokeh/dev This is almost passing tests now, with an up to date toolchain. One caveat: 

New versions of conda build do not seem to support building "in place" the way we did before. So I have updated `.travis.yml` to install build deps and build in place. I still thing some test somewhere should run conda build, just to keep on the notion that building it does in fact work. Does this seem OK? @msarahan may have updates that allow for building in place soon, then we could go back to how things were. 

@mattpap the only tests failing are the server and notebook examples tests. I am tkaing a look but I may need your help.

**Comment by bryevdv** [comment]
@mattpap @birdsarah @damianavila @bokeh/dev Update on the caveat:

The `conda.recipe` does sort of now build "in place" -- in particular it does not really build at all, it assumes that `--build-js` has previously been run, and just calls `setup.py` with `--install-js`. This will be considerably faster than re-installing NPM for every python version build. Until we switch to (new-style) no-arch packages, I think this is a reasonable route, but I would appreciate feedback .

I'd like to merge this soon and test out the machinery on a dev build. I have tested all the build functions, RC credentials and uploading functions as best I can locally.

File: MANIFEST.in [hunk 1]
```diff
@@ -1,6 +1,10 @@
 global-exclude .git*
+include CHANGELOG
+include classifiers.txt
 include LICENSE.txt
+include setup.cfg
 include versioneer.py
+include _setup_support.py
```
Comment (bryevdv) [inline MANIFEST.in]: These need to be added here to be included inside sdist tarballs (they are needed to build from sdist)

**Comment by bryevdv** [comment]
I uncovered some problems with the handling of sdists and some other minor issues. I'd like to merge this tomorrow and use it to try out a dev build.

**Comment by bryevdv** [comment]
There are some further improvements that could be made (e.g. the conda build mocking) would be handled more cleanly by making a custom command, but I think this is enough for now.

File: tests/test_code_quality.py [hunk 2]
```diff
@@ -93,9 +93,9 @@ def check_files(files, file_exclusions=None):
 
             test(fname)
 
-    check_files(["setup.py"])
+    check_files(["setup.py", "_setup_support.py"])
     check_tree('bin',          ['*'])
-    check_tree('bokeh',        ['*.py', '*.html', '*.js'], ["server/static"], ["__conda_version__.py"])
+    check_tree('bokeh',        ['*.py', '*.html', '*.js'], ["server/static"])
```
Comment (mattpap) [inline tests/test_code_quality.py]: Those changes won't be needed after PR #5751. We can resolve merge conflicts either way.

**Comment by mattpap** [comment]
@bryevdv, I presume it's not directly connected to this PR, but is there a way to create a local test environment based on `conda.recipe/meta.yaml`, so that I don't have to parse YAML manually? I would image something like `conda env create --use=test conda.recipe/meta.yaml`?

**Comment by mattpap** [comment]
LGTM, though I think I don't have too much expertise with this to discuss particular points, but I like very much how the overall design got simplified.

**Comment by msarahan** [comment]
@mattpap what you want with meta.yaml unfortunately doesn't exist right now, but I'd like to add it soon.  Probably in the next month or so.

**Comment by mattpap** [comment]
@msarahan, thanks, looking forward to this. Is there an issue for this?

**Comment by bryevdv** [comment]
@mattpap great thanks for taking a look. I will go ahead and merge and test out a dev build

**Comment by msarahan** [comment]
@mattpap there wasn't one, I don't think, but I have made this: https://github.com/conda/conda-build/issues/1677

**Comment by bryevdv** [comment]
Restoring the branch, it didn't take long to fall over after merged to master. May need to revert.

**Comment by bryevdv** [comment]
`0.12.5.dev1` is not a valid version for npm. It absolutely refuses to install with that version in `package.json` which sucks. Alot. Do we:

* abandon pep440 compliance for python?
* abandon matching version numbers python vs js?
* adopt a different scheme for dev builds that works for both: `0.12.5a1` ?

**Comment by mattpap** [comment]
> adopt a different scheme for dev builds that works for both: 0.12.5a1 ?

This seems to suck the least.

**Comment by canavandl** [comment]
I agree ^

**Comment by bryevdv** [comment]
it looks like there is a chance `-dev2` will work, investigating that first.

**Comment by bryevdv** [comment]
After some discussion with others, it seems like minor deviation from pep440 with our existing scheme `0.12.5dev2` will be OK, if occasionally annoying.

**Pull request merged**